results/results.go

```diff
@@ -0,0 +1,171 @@
+				continue // only 1 _ separator at a time
+			}
+			last = '_'
+		}
+		base += string(last)
+	}
+	if last == '_' {
+		base = base[:len(base)-1]
+	}
+	if len(base) > 64 {
+		return base[:64]
+	}
+	return base
+}
+
+// HasRunnerResult is the interface implictly implemented by HTTPRunnerResults
+// and GrpcRunnerResults so the common results can ge extracted irrespective
+// of the type.
+type HasRunnerResult interface {
+	Result() *RunnerResults
+}
+
+// Result returns the common RunnerResults.
+func (r *RunnerResults) Result() *RunnerResults {
+	return r
+}
+
+// SaveJSON saves the result as json to the provided file
+//
+func SaveJSON(res HasRunnerResult, fileName string) (int, error) {
+	var j []byte
+	j, err := json.MarshalIndent(res, "", "  ")
+	if err != nil {
+		log.Errf("Unable to json serialize result: %v", err)
+		return -1, err
+	}
+	var f *os.File
+	if fileName == "-" {
+		f = os.Stdout
+		fileName = "stdout"
+	} else {
+		f, err = os.Create(fileName)
+		if err != nil {
+			log.Errf("Unable to create %s: %v", fileName, err)
+			return -1, err
+		}
+	}
+	n, err := f.Write(append(j, '\n'))
+	if err != nil {
+		log.Errf("Unable to write json to %s: %v", fileName, err)
+		return -1, err
+	}
+	if f != os.Stdout {
+		err := f.Close()
+		if err != nil {
+			log.Errf("Close error for %s: %v", fileName, err)
+			return -1, err
+		}
+	}
+	return n, nil
+}
+
+func ParseArgsForResultFiles(args []string) ([]string, error) {
+	res := make([]string, 0, len(args))
+	for _, file := range args {
+		file = strings.TrimSpace(file)
+		if len(file) == 0 {
+			continue
+		}
+		if strings.HasSuffix(file, ".json") {
+			res = append(res, file)
+		}
+	}
+	log.LogVf("Identified %v as result files", res)
+	return res, nil
+}
+
+func LoadResultFiles(files[] string) ([]*RunnerResults, error) {
+	res := make([]*RunnerResults, 0, len(files))
+	for _, file := range files {
+		raw, err := ioutil.ReadFile(file)
+		if err != nil {
+			log.Errf("Error reading file %s", file)
+			return nil, err
+		}
+		result := RunnerResults{}
+		json.Unmarshal(raw, &result)
+		res = append(res, &result)
+	}
+	return res, nil
+}
+
+func MergeRunnerResults(results []*RunnerResults) {
+	// TODO determine if results are mergeable
+	// TODO specify type of results in the json file metadata
+	// TODO check for len(files) > 2 and print usage etc
+
+	// iterate and merge each sequentially
+	mergedResults := results[0]
+	for _, result := range results[1:] {
+		// merge Labels
+		// merge StartTime (earliest wins)
+		if mergedResults.StartTime.After(result.StartTime) {
+			mergedResults.StartTime = result.StartTime
+		}
+		// merge RequestedQPS
+		// merge RequestedDuration
+		// ActualQPS
+		// ActualDuration
+		// merge NumThreads
+		mergedResults.NumThreads += result.NumThreads
+		// merge Version
+		// merge HistogramData
+		mergedResults.DurationHistogram = stats.MergeHistograms(mergedResults.DurationHistogram.Histogram(), result.DurationHistogram.Histogram()).Export()
```

Comment:
this is a lot of redundant conversions between Histogram and HistogramData. However, merging HistogramData seems more complicated due to the possible differences in the percentiles or buckets?

Reply:
maybe an initial version could only merge histograms with same settings; on the other hand the fun part is to merge the ones that are different. the function on histogram isn't perfect (and not used really anywhere but in tests)